From: Armin Novak <armin.novak@thincast.com>
Date: Mon, 28 Mar 2022 10:32:08 +0200
Subject: Fixed NTLM flag handling

* NTLMSSP_NEGOTIATE_WORKSTATION_SUPPLIED and NTLMSSP_NEGOTIATE_DOMAIN_SUPPLIED
* Problem with clients not supporting NTLMSSP_NEGOTIATE_LM_KEY
* Improve logging

(cherry picked from commit 3c657d58cd0e7c2398b5488902ecf77bdb2eb8d8)
(cherry picked from commit 269ebb8cada3fef6b6cabfa5a1ea98ff4abbb99e)
---
 winpr/libwinpr/sspi/NTLM/ntlm.c         | 35 +++++++++++++++++++--------------
 winpr/libwinpr/sspi/NTLM/ntlm_message.c | 26 +++++++++++++++++-------
 2 files changed, 39 insertions(+), 22 deletions(-)

diff --git a/winpr/libwinpr/sspi/NTLM/ntlm.c b/winpr/libwinpr/sspi/NTLM/ntlm.c
index c0af65b..e18fea2 100644
--- a/winpr/libwinpr/sspi/NTLM/ntlm.c
+++ b/winpr/libwinpr/sspi/NTLM/ntlm.c
@@ -1289,26 +1289,31 @@ char* ntlm_negotiate_flags_string(char* buffer, size_t size, UINT32 flags)
 	if (!buffer || (size == 0))
 		return buffer;
 
-	for (x = 0; x < 31; x++)
-	{
-		size_t len = strnlen(buffer, size);
-		if (flags & x)
+	snprintf(buffer, size, "[0x%08" PRIx32 "] ", flags);
+
+		for (x = 0; x < 31; x++)
 		{
-			const char* str = ntlm_get_negotiate_string(1 << x);
-			size_t flen = strlen(str);
-			if (len > 0)
+			const UINT32 mask = 1 << x;
+			size_t len = strnlen(buffer, size);
+			if (flags & mask)
 			{
-				if (size - len < 1)
+				const char* str = ntlm_get_negotiate_string(mask);
+				const size_t flen = strlen(str);
+
+				if ((len > 0) && (buffer[len - 1] != ' '))
+				{
+					if (size - len < 1)
+						break;
+					strcat(buffer, "|");
+					len++;
+				}
+
+				if (size - len < flen)
 					break;
-				strcat(buffer, "|");
-				len++;
+				strcat(buffer, str);
 			}
-
-			if (size - len < flen)
-				break;
-			strcat(buffer, str);
 		}
-	}
+
 	return buffer;
 }
 
diff --git a/winpr/libwinpr/sspi/NTLM/ntlm_message.c b/winpr/libwinpr/sspi/NTLM/ntlm_message.c
index 7f3779d..442b9fe 100644
--- a/winpr/libwinpr/sspi/NTLM/ntlm_message.c
+++ b/winpr/libwinpr/sspi/NTLM/ntlm_message.c
@@ -417,6 +417,7 @@ static void ntlm_free_message_fields_buffer(NTLM_MESSAGE_FIELDS* fields)
 static BOOL ntlm_read_negotiate_flags(wStream* s, UINT32* flags, UINT32 required, const char* name)
 {
 	UINT32 NegotiateFlags = 0;
+	char buffer[1024] = { 0 };
 	WINPR_ASSERT(s);
 	WINPR_ASSERT(flags);
 	WINPR_ASSERT(name);
@@ -436,12 +437,16 @@ static BOOL ntlm_read_negotiate_flags(wStream* s, UINT32* flags, UINT32 required
 		         name, NegotiateFlags, required);
 		return FALSE;
 	}
+
+	WLog_DBG(TAG, "Read flags %s",
+	         ntlm_negotiate_flags_string(buffer, ARRAYSIZE(buffer), NegotiateFlags));
 	*flags = NegotiateFlags;
 	return TRUE;
 }
 
 static BOOL ntlm_write_negotiate_flags(wStream* s, UINT32 flags, const char* name)
 {
+	char buffer[1024] = { 0 };
 	WINPR_ASSERT(s);
 	WINPR_ASSERT(name);
 
@@ -452,6 +457,7 @@ static BOOL ntlm_write_negotiate_flags(wStream* s, UINT32 flags, const char* nam
 		return FALSE;
 	}
 
+	WLog_DBG(TAG, "Write flags %s", ntlm_negotiate_flags_string(buffer, ARRAYSIZE(buffer), flags));
 	Stream_Write_UINT32(s, flags); /* NegotiateFlags (4 bytes) */
 	return TRUE;
 }
@@ -565,6 +571,12 @@ SECURITY_STATUS ntlm_read_NegotiateMessage(NTLM_CONTEXT* context, PSecBuffer buf
 			return SEC_E_INVALID_TOKEN;
 	}
 
+		if (!ntlm_read_message_fields_buffer(s, &message->DomainName))
+			return SEC_E_INVALID_TOKEN;
+
+		if (!ntlm_read_message_fields_buffer(s, &message->Workstation))
+			return SEC_E_INVALID_TOKEN;
+
 	length = Stream_GetPosition(s);
 	WINPR_ASSERT(length <= ULONG_MAX);
 	buffer->cbBuffer = (ULONG)length;
@@ -1020,11 +1032,8 @@ SECURITY_STATUS ntlm_read_AuthenticateMessage(NTLM_CONTEXT* context, PSecBuffer
 	if (!ntlm_read_message_fields_buffer(s, &(message->UserName))) /* UserName */
 		goto fail;
 
-	if (message->NegotiateFlags & NTLMSSP_NEGOTIATE_WORKSTATION_SUPPLIED)
-	{
-		if (!ntlm_read_message_fields_buffer(s, &(message->Workstation))) /* Workstation */
-			goto fail;
-	}
+	if (!ntlm_read_message_fields_buffer(s, &(message->Workstation))) /* Workstation */
+		goto fail;
 
 	if (message->NegotiateFlags & NTLMSSP_NEGOTIATE_LM_KEY)
 	{
@@ -1364,8 +1373,11 @@ SECURITY_STATUS ntlm_server_AuthenticateComplete(NTLM_CONTEXT* context)
 	if (AvFlags)
 		Data_Read_UINT32(ntlm_av_pair_get_value_pointer(AvFlags), flags);
 
-	if (!ntlm_compute_lm_v2_response(context)) /* LmChallengeResponse */
-		return SEC_E_INTERNAL_ERROR;
+	if (context->NegotiateFlags & NTLMSSP_NEGOTIATE_LM_KEY)
+	{
+		if (!ntlm_compute_lm_v2_response(context)) /* LmChallengeResponse */
+			return SEC_E_INTERNAL_ERROR;
+	}
 
 	if (!ntlm_compute_ntlm_v2_response(context)) /* NtChallengeResponse */
 		return SEC_E_INTERNAL_ERROR;
